sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/GpuBloomFilter.scala

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*** spark-rapids-shim-json-lines
+{"spark": "330"}
+{"spark": "330cdh"}
+{"spark": "330db"}
+{"spark": "331"}
+{"spark": "332"}
+{"spark": "332db"}
+{"spark": "333"}
+{"spark": "340"}
+{"spark": "341"}
+spark-rapids-shim-json-lines ***/
+package com.nvidia.spark.rapids
+
+import java.io.DataInputStream
+
+import ai.rapids.cudf.{BaseDeviceMemoryBuffer, ColumnVector, Cuda, DeviceMemoryBuffer, DType, HostMemoryBuffer}
+import com.nvidia.spark.rapids.Arm.{closeOnExcept, withResource}
+import com.nvidia.spark.rapids.jni.BloomFilter
+
+import org.apache.spark.sql.types.BinaryType
+
+/** GPU version of Spark's BloomFilterImpl */
+class GpuBloomFilter(val numHashes: Int, buffer: DeviceMemoryBuffer) extends AutoCloseable {
+  private val spillableBuffer = SpillableBuffer(buffer, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+  private val numFilterBits = buffer.getLength * 8
+
+  /**
+   * Given an input column of longs, return a boolean column with the same row count where each
+   * output row indicates whether the corresponding input row may have been placed into this
+   * Bloom filter. A false value indicates definitively that the value was not placed in the filter.
+   */
+  def mightContainLong(col: ColumnVector): ColumnVector = {
+    require(col.getType == DType.INT64, s"expected longs, got ${col.getType}")
+    withResource(spillableBuffer.getDeviceBuffer()) { buffer =>
+      buffer.incRefCount()
+      withResource(new BloomFilter(numHashes, numFilterBits, buffer)) { filter =>
+        filter.probe(col)
+      }
+    }
+  }
+
+  override def close(): Unit = {
+    spillableBuffer.close()
+  }
+}
+
+object GpuBloomFilter {
+  // Spark serializes their bloom filters in a specific format, see BloomFilterImpl.readFrom.
+  // Data is written via DataOutputStream, so everything is big-endian.
+  // Byte Offset  Size  Description
+  // 0            4     Version ID (see Spark's BloomFilter.Version)
+  // 4            4     Number of hash functions
+  // 8            4     Number of longs, N
+  // 12           N*8   Bloom filter data buffer as longs
+  private val HEADER_SIZE = 12
+
+  // version numbers from BloomFilter.Version enum
+  private val VERSION_V1 = 1
+
+  def apply(s: GpuScalar): GpuBloomFilter = {
+    if (s == null) {
+      null
+    } else {
+      s.dataType match {
+        case BinaryType =>
+          withResource(s.getBase.getListAsColumnView) { childView =>
+            require(childView.getType == DType.UINT8, s"expected UINT8 got ${childView.getType}")
+            deserialize(childView.getData)
+          }
+        case t => throw new IllegalArgumentException(s"Expected binary scalar, found $t")
```

Comment:
So in the check bloom filter code we say that we support NullType, but if it shows up here we will throw an exception.